demos/ota/ota_demo_core_mqtt/ota_demo_core_mqtt.c

```diff
@@ -0,0 +1,1354 @@
+    pSubscriptionList[ 0 ].qos = qos;
+    pSubscriptionList[ 0 ].pTopicFilter = pTopicFilter;
+    pSubscriptionList[ 0 ].topicFilterLength = topicFilterLength;
+
+    /* Send SUBSCRIBE packet. */
+    mqttStatus = MQTT_Subscribe( pMqttContext,
+                                 pSubscriptionList,
+                                 sizeof( pSubscriptionList ) / sizeof( MQTTSubscribeInfo_t ),
+                                 MQTT_GetPacketId( pMqttContext ) );
+
+    if( mqttStatus != MQTTSuccess )
+    {
+        LogError( ( "Failed to send SUBSCRIBE packet to broker with error = %u.",
+                    mqttStatus ) );
+
+        otaRet = OtaMqttSubscribeFailed;
+    }
+    else
+    {
+        LogInfo( ( "SUBSCRIBE topic %.*s to broker.\n\n",
+                   topicFilterLength,
+                   pTopicFilter ) );
+    }
+
+    /* Register callback to subscription manager. */
+    SubscriptionManager_RegisterCallback( pTopicFilter, topicFilterLength, callback );
+
+    return otaRet;
+}
+
+/*-----------------------------------------------------------*/
+
+static OtaMqttStatus_t mqttPublish( const char * const pacTopic,
+                                    uint16_t topicLen,
+                                    const char * pMsg,
+                                    uint32_t msgSize,
+                                    uint8_t qos )
+{
+    OtaMqttStatus_t otaRet = OtaMqttSuccess;
+
+    MQTTStatus_t mqttStatus = MQTTBadParameter;
+    MQTTPublishInfo_t publishInfo;
+    MQTTContext_t * pMqttContext = &mqttContext;
+
+    /* Set the required publish parameters. */
+    publishInfo.pTopicName = pacTopic;
+    publishInfo.topicNameLength = topicLen;
+    publishInfo.qos = qos;
+    publishInfo.pPayload = pMsg;
+    publishInfo.payloadLength = msgSize;
+
+    mqttStatus = MQTT_Publish( pMqttContext,
+                               &publishInfo,
+                               MQTT_GetPacketId( pMqttContext ) );
+
+    if( mqttStatus != MQTTSuccess )
+    {
+        LogError( ( "Failed to send PUBLISH packet to broker with error = %u.", mqttStatus ) );
+
+        otaRet = OtaMqttPublishFailed;
+    }
+    else
+    {
+        LogInfo( ( "Sent PUBLISH packet to broker %.*s to broker.\n\n",
+                   topicLen,
+                   pacTopic ) );
+    }
+
+    return otaRet;
+}
+
+/*-----------------------------------------------------------*/
+
+static OtaMqttStatus_t mqttUnsubscribe( const char * pTopicFilter,
+                                        uint16_t topicFilterLength,
+                                        uint8_t qos )
+{
+    OtaMqttStatus_t otaRet = OtaMqttSuccess;
+    MQTTStatus_t mqttStatus;
+
+    MQTTSubscribeInfo_t pSubscriptionList[ 1 ];
+    MQTTContext_t * pMqttContext = &mqttContext;
+
+    /* Start with everything at 0. */
+    ( void ) memset( ( void * ) pSubscriptionList, 0x00, sizeof( pSubscriptionList ) );
+
+    /* Set the QoS , topic and topic length. */
+    pSubscriptionList[ 0 ].qos = qos;
```

Comment:
QoS is not required for sending an unsubscribe packet to the broker. 
It would be out of scope for this PR, but the OTA interface for unsubscribe can consider removing the QoS parameter in the future

Reply:
Agreed, can fix in a later PR.